8.4: Document all new DOM classes and methods - part 1

nielsdos opened the pull request:
Very WIP, not ready for review yet.

TODO:
- [x] Update other extensions that link into DOM
- [x] Update constants
- [x] Finish class skeletons (class page, properties, ...)
- [x] Document the NO_DEFAULT_NS constant (with caution)
- [ ] Document methods (NOTE: splitting PR because this is huge already)
- [x] Update note about encoding
- [ ] ~~Create examples of array indexing in NamedNodeMap ?~~ Let's do this separately
- [ ] ~~Add a note on top of the dom book to guide users to the new classes?~~ Do this later once the docs are completed
- [ ] ~~Explain xpath subtleties~~ Separate PR
- [x] Add note to XMLDocument properties that you should rely on the LIBXML_ flags now

nielsdos commented:
@Girgias I've finished with some basics. Most notably the class synopses are generated and the properties etc are all filled in. I've used xinclude quite a bit. I also fixed up some of the old DOM class descriptions.
It's probably best to already review this as-is, even though it's not finished and most notably the methods are missing.
This won't compile out of the box because of the entities.whatever.xml files not being generated because I haven't filled in the method stuff yet, but here's a zip that you can extract in `reference/dom/dom` to make it build:
[entities.dom.zip](https://github.com/user-attachments/files/17970335/entities.dom.zip)

Girgias commented on reference/dom/constants.xml:
```diff
@@ -277,26 +277,26 @@
      </row>
      <row xml:id="constant.domstring-size-err">
       <entry>
-       <constant>DOMSTRING_SIZE_ERR</constant>
+       <constant>DOMSTRING_SIZE_ERR</constant> / <constant>Dom\STRING_SIZE_ERR</constant>
```
The issue I can see for this, is that automatic linking for `<constant>Dom\STRING_SIZE_ERR</constant>` is not going to work because the row has no corresponding XML ID.

Can an XML element have 2 IDs? If yes that would fix it.

nielsdos replied:
> Can an XML element have 2 IDs? If yes that would fix it.

No, but I don't think this is ever linked to directly :thinking:

Girgias commented on reference/dom/dom/dom-attr.xml:
```diff
@@ -0,0 +1,163 @@
+    <varlistentry xml:id="dom-attr.props.namespaceuri">
+     <term><varname>namespaceURI</varname></term>
+     <listitem>
+      <simpara>The namespace <abbrev>URI</abbrev> of the attribute.</simpara>
```
I don't know if we actually have any support for `<abbrev>`, but in any case URI is an acronym, not an abbreviation.

Please use:
```suggestion
      <simpara>The namespace <acronym>URI</acronym> of the attribute.</simpara>
```

nielsdos replied:
Cool, I consistently did this wrong :grimacing: . I think it is because in my native language we misuse the same word in speech sometimes for both concepts...

Girgias replied:
I did need to double check the definitions myself, so I get it :D

Girgias commented on reference/dom/dom/dom-attr.xml:
```diff
@@ -0,0 +1,163 @@
+      <simpara>The element that contains the attribute or &null;.</simpara>
+     </listitem>
+    </varlistentry>
+    <varlistentry xml:id="dom-attr.props.specified">
+     <term><varname>specified</varname></term>
+     <listitem>
+      <simpara>Legacy option, always is &true;.</simpara>
+     </listitem>
+    </varlistentry>
```
Should this have been added to the new classes actually?

nielsdos replied:
Yesn't. This is still defined by spec and is here also so this class can be used as a drop-in replacement.

Girgias replied:
ACK

Girgias commented on reference/dom/dom/dom-document.xml:
```diff
@@ -0,0 +1,250 @@
+      </simpara>
+     </listitem>
+    </varlistentry>
+    <varlistentry xml:id="dom-document.props.inputencoding">
+     <term><varname>inputEncoding</varname></term>
+     <listitem>
+      <simpara>Legacy alias for <varname>characterSet</varname>.</simpara>
+     </listitem>
+    </varlistentry>
+    <varlistentry xml:id="dom-document.props.charset">
+     <term><varname>charset</varname></term>
+     <listitem>
+      <simpara>Legacy alias for <varname>characterSet</varname>.</simpara>
+     </listitem>
+    </varlistentry>
```
Ditto, should those have been added on the new classes?

nielsdos replied:
Defined by spec + BC. Although at some point we could be splitting inputEncoding from the rest which would make some sense in PHP and not make sense in the JS world.

Girgias replied:
ACK.

Girgias commented on reference/dom/dom/dom-document.xml:
```diff
@@ -0,0 +1,250 @@
+     <term><varname>body</varname></term>
+     <listitem>
+      <simpara>
+       The first child of the html element that is either a body tag or a
+       frameset tag.
```
Should a `<literal>` tag be used to refer to HTML elements/tags?

nielsdos replied:
good point!

Girgias commented on reference/dom/dom/dom-dtdnamednodemap.xml:
```diff
@@ -0,0 +1,67 @@
+   &reftitle.intro;
+   <simpara>
+    Represents a named node map for entities and notation nodes of the
+    <abbrev>DTD</abbrev>.
```
DTD is an acronym, not an abbreviation

Girgias commented on reference/dom/dom/dom-htmldocument.xml:
```diff
@@ -0,0 +1,67 @@
+  <section xml:id="dom-htmldocument.intro">
+   &reftitle.intro;
+   <simpara>
+    Represents an <abbrev>HTML</abbrev> document.
```
Ditto acronym

Girgias commented on reference/dom/dom/dom-htmlelement.xml:
```diff
@@ -0,0 +1,59 @@
+  <section xml:id="dom-htmlelement.intro">
+   &reftitle.intro;
+   <simpara>
+    Represents an element in the <abbrev>HTML</abbrev> namespace.
```
Ditto

Girgias commented on reference/dom/dom/dom-node.xml:
```diff
@@ -0,0 +1,310 @@
+     <listitem>
+      <simpara>
+       The <classname>Dom\Document</classname> object associated with
+       this node, or null if this node is a document.
```
```suggestion
       this node, or &null; if this node is a document.
```

Girgias commented on reference/dom/dom/dom-xmldocument.xml:
```diff
@@ -0,0 +1,117 @@
+  <section xml:id="dom-xmldocument.intro">
+   &reftitle.intro;
+   <simpara>
+    Represents an <abbrev>XML</abbrev> document.
```
Ditto acronym

Girgias commented on reference/dom/domcharacterdata.xml:
```diff
@@ -9,10 +9,10 @@
    &reftitle.intro;
-   <para>
+   <simpara>
     Represents nodes with character data. No nodes directly correspond to
     this class, but other nodes do inherit from it.
-   </para>
+   </simpara>
```
I wouldn't change this, yes it is not according to our modern style guide, but not sure it is worth bothering translations.

Girgias commented on reference/dom/domcharacterdata/deletedata.xml:
```diff
@@ -59,7 +59,7 @@
       <para>
        Raised if <parameter>offset</parameter> is negative or greater than the
-       number of 16-bit units in data, or if <parameter>count</parameter> is
+       number of UTF-8 codepoints in data, or if <parameter>count</parameter> is
```
is "data" referring to a property?

nielsdos replied:
Yes

Girgias commented on reference/dom/domcharacterdata/replacedata.xml:
```diff
@@ -68,7 +68,7 @@
       <para>
        Raised if <parameter>offset</parameter> is negative or greater than the
-       number of 16-bit units in data, or if <parameter>count</parameter> is
+       number of UTF-8 codepoints in data, or if <parameter>count</parameter> is
```
Ditto

Edit: this seems to be a "consistent" issue, so I'd relegate this to a later PR.

Girgias commented on reference/dom/domcomment.xml:
```diff
@@ -9,10 +9,10 @@
    &reftitle.intro;
-   <para>
+   <simpara>
     Represents comment nodes, characters delimited by <code>&lt;!--</code>
     and <code>--&gt;</code>.
-   </para>
+   </simpara>
```
Ditto, not sure it is worth

Girgias commented on reference/dom/domdocument.xml:
```diff
@@ -9,10 +9,10 @@
    &reftitle.intro;
-   <para>
+   <simpara>
     Represents an entire HTML or XML document; serves as the root of the
     document tree.
-   </para>
+   </simpara>
```
Ditto this is the only change

Girgias commented on reference/dom/domdocumenttype.xml:
```diff
@@ -102,15 +102,15 @@
       <para>
        The system identifier of the external subset. This may be an
-       absolute URI or not.
+       absolute <abbrev>URI</abbrev> or not.
```
Changes from this whole file, those are acronyms.

Girgias commented on reference/dom/domentity.xml:
```diff
@@ -9,9 +9,9 @@
   <section xml:id="domentity.intro">
    &reftitle.intro;
-   <para>
+   <simpara>
     This interface represents a known entity, either parsed or unparsed, in an XML document.
-   </para>
+   </simpara>
```
Ditto not sure it is worth to change this

Girgias commented on reference/xsl/xsltprocessor/transformtoxml.xml:
```diff
@@ -24,8 +24,8 @@
       <para>
-       The <classname>DOMDocument</classname> or <classname>SimpleXMLElement</classname> object to
-       be transformed.
+       The <classname>Dom\Document</classname>, <classname>DOMDocument</classname>, <classname>SimpleXMLElement</classname> or libxml-compatible
+       object to be transformed.
```
You could XInclude this from the toUrl method

Girgias reviewed (commented):
Didn't test locally but this a pass across the whole PR

nielsdos commented:
Thanks for the initial review. I pushed fixup commits so they can be validated individually, I'll later rebase this to merge the fixups with the original commits.

Girgias commented on reference/dom/dom/dom-entity.xml:
```diff
@@ -7,7 +7,7 @@
   <section xml:id="dom-entity.intro">
    &reftitle.intro;
-   <xi:include xpointer="xmlns(db=http://docbook.org/ns/docbook) xpointer(id('domentity.intro')/db:simpara)">
+   <xi:include xpointer="xmlns(db=http://docbook.org/ns/docbook) xpointer(id('domentity.intro')/db:para)">
```
Why not just Xinclude the whole introduction section instead of just including the `para`? (same for some of the other pages)

Girgias replied:
Forgot about this?

Girgias commented on reference/dom/dom/dom-comment.xml:
```diff
@@ -0,0 +1,67 @@
+
+  <section role="seealso">
+   &reftitle.seealso;
+   <para>
+    <simplelist>
+     <member><link xlink:href="&url.spec.dom.living.comment;">WHATWG specification of Comment</link></member>
+    </simplelist>
+   </para>
```
No need for the wrapping para tag here (haven't checked all of the files, but it might be in others?)

Girgias commented on reference/dom/dom/dom-text.xml:
```diff
@@ -0,0 +1,83 @@
+
+  <section xml:id="dom-text.intro">
+   &reftitle.intro;
+   <para>
+    The <classname>Dom\Text</classname> class inherits from
+    <classname>Dom\CharacterData</classname> and represents a text node.
+   </para>
```
New file, so `<simpara>`

Girgias commented on reference/dom/dom/attr/isid.xml:
```diff
@@ -0,0 +1,74 @@
+  </methodsynopsis>
+  <xi:include xpointer="xmlns(db=http://docbook.org/ns/docbook) xpointer(id('domattr.isid')/db:refsect1[@role='description']/db:para[1])"/>
+  <simpara>
+   According to the DOM standard this requires a DTD which defines the
+   attribute ID to be of type ID. You need to validate your document by
+   passing <constant>LIBXML_DTDVALID</constant> to the parse method.
```
Personalization is used here.

Girgias commented on reference/dom/dom/characterdata/before.xml:
```diff
@@ -0,0 +1,78 @@
+  &reftitle.examples;
+  <example xml:id="dom-characterdata.before.example.basic">
+   <title><methodname>Dom\CharacterData::before</methodname> example</title>
+   <para>
+    Adds nodes before the character data.
+   </para>
```
Simpara

Girgias commented on reference/dom/dom/characterdata/remove.xml:
```diff
@@ -0,0 +1,80 @@
+  &reftitle.examples;
+  <example xml:id="dom-characterdata.remove.example.basic">
+   <title><methodname>Dom\CharacterData::remove</methodname> example</title>
+   <para>
+    Removes the character data.
+   </para>
```
simpara

Girgias commented on reference/dom/dom/characterdata/replacewith.xml:
```diff
@@ -0,0 +1,80 @@
+  &reftitle.examples;
+  <example xml:id="dom-characterdata.replacewith.example.basic">
+   <title><methodname>Dom\CharacterData::replaceWith</methodname> example</title>
+   <para>
+    Replaces the character data with new nodes.
+   </para>
```
simpara

Girgias commented on reference/dom/dom/attr/rename.xml:
```diff
@@ -0,0 +1,176 @@
+
+ <refsect1 role="returnvalues">
+  &reftitle.returnvalues;
+  <para>
+   &return.void;
+  </para>
+ </refsect1>
```
simpara

nielsdos commented:
@Girgias I brought this PR to a compilable state as we discussed. I addressed your review comments and rebased on master. The fixes for your review comments start at "[fixup! comment](https://github.com/php/doc-en/pull/4212/commits/b74ee415fb2e091a0214e03721de99336c66724d)".
It'd be great to merge this then and create separate PRs per class for the methods that are not documented yet.

Girgias commented on reference/dom/constants.xml:
```diff
@@ -241,6 +241,39 @@
+   <tgroup cols="3">
+    <thead>
+     <row>
+      <entry>Constant</entry>
+      <entry>Description</entry>
```
Please use the entities for these.

nielsdos replied:
Do you mean for the "constant" & "description" header row?

Girgias replied:
Yes, should be `&Constants;` and `&Description;`

Girgias commented on reference/dom/dom/dom-xmldocument.xml:
```diff
@@ -82,6 +82,14 @@
+   <note>
+    <simpara>
+     While the <classname>DOMDocument</classname> class allows setting certain
+     properties to influence parser behaviour, this class requires you to pass
+     <constant>LIBXML_<replaceable>*</replaceable></constant> options to the
+     parser methods.
```
Usage of "you"

Girgias commented on language-snippets.ent:
```diff
@@ -1648,30 +1648,35 @@ it is inserted with (e.g.) <function xmlns="http://docbook.org/ns/docbook">DOMNo
 
 <!ENTITY dom.malformederror '<para xmlns="http://docbook.org/ns/docbook">While malformed HTML should load successfully, this function may generate <constant>E_WARNING</constant> errors when it encounters bad markup. <link linkend="function.libxml-use-internal-errors">libxml&apos;s error handling functions</link> may be used to handle these errors.</para>'>
 <!ENTITY dom.note.utf8 '<note xmlns="http://docbook.org/ns/docbook"><para>The DOM extension uses UTF-8 encoding. Use <function>mb_convert_encoding</function>, <methodname>UConverter::transcode</methodname>, or <function>iconv</function> to handle other encodings.</para></note>'>
+<!ENTITY dom.note.modern.utf8 '<note xmlns="http://docbook.org/ns/docbook"><simpara>The DOM extension uses UTF-8 encoding when working with methods or properties. The parser methods auto-detect the encoding or allow the caller to specify an encoding.</simpara></note>'>
```
Just to not be compressed
```suggestion
<!ENTITY dom.note.modern.utf8 '<note xmlns="http://docbook.org/ns/docbook">
 <simpara>
  The DOM extension uses UTF-8 encoding when working with methods or properties.
  The parser methods auto-detect the encoding or allow the caller to specify an encoding.
 </simpara>
</note>'>
```

Girgias commented on reference/dom/dom/attr/rename.xml:
```diff
@@ -0,0 +1,176 @@
+
+ <refsect1 role="parameters">
+  &reftitle.parameters;
+  <para>
+   <variablelist>
+    <varlistentry>
+     <term><parameter>namespaceURI</parameter></term>
+     <listitem>
+      <simpara>
+       The new namespace <acronym>URI</acronym> of the attribute.
+      </simpara>
+     </listitem>
+    </varlistentry>
+    <varlistentry>
+     <term><parameter>qualifiedName</parameter></term>
+     <listitem>
+      <simpara>
+       The new qualified name of the attribute.
+      </simpara>
+     </listitem>
+    </varlistentry>
+   </variablelist>
+  </para>
```
Useless para wrapping tag

Girgias reviewed (approved):
LGTM, thank you!
